nomenclature/region_mapping_models.py

```diff
@@ -104,3 +107,45 @@ def create_from_region_mapping(cls, file: Union[Path, str]):
                 )
             mapping_input["common_regions"] = common_region_list
         return cls(**mapping_input)
+
+
+class RegionProcessor(Mapping):
+    """Thin wrapper around a dict to hold a collection of region mappings"""
+
+    def __init__(
+        self, mapping_path: Path, dsd: Optional[DataStructureDefinition] = None
+    ) -> None:
+        if not isinstance(mapping_path, Path):
+            raise TypeError(f"mapping_path must be Path, is {type(mapping_path)}")
+        if not mapping_path.is_dir():
+            raise ValueError("mapping_path must point to a directory")
+        self._ra_mappings: Dict[str, RegionAggregationMapping] = {}
+        self.parse_mappings(mapping_path)
+        if dsd is not None:
+            self.validate_regions(dsd)
+
+    def parse_mappings(self, mapping_path: Path) -> None:
+        for file in mapping_path.glob("**/*.yaml"):
+            ram = RegionAggregationMapping.create_from_region_mapping(file)
+            self[ram.model] = ram
+
+    def validate_regions(self, dsd: DataStructureDefinition) -> None:
+        # not implemented for now
+        pass
+
+    def __setitem__(self, key, value):
+        if key in self._ra_mappings:
+            raise KeyError(f"Duplicate mapping for model {key}")
+        self._ra_mappings[key] = value
+
+    def __getitem__(self, k):
```

Comment:
As I'm implementing this right now I have realized a problem with using `_mapping` as an internal private attribute. 

Pydantic doesn't really allow for that (here's the [pydantic issue](https://github.com/samuelcolvin/pydantic/issues/288) detailing the discussion about it). To be precise private attributes are allowed but they need to be set extra by overwriting the constructor. There is an option for private attributes [example from the pydantic documentation](https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes) but it requires a modified `__init__`. In principle that's not a problem but it our case I would argue that `RegionProcessor` should not have any other attributes besides this dictionary which holds the mappings. Both the `DataStructureDefinition` and the `IamDataFrame`(s) that we want to use for validation and aggregation respectively should be function inputs and not part of the class. This would mean that we would initiate `RegionProcessor` with a single private attribute and not use the normal `pydanitic.BaseModel.__init__` at all. At this point I would argue pydantic is the wrong choice.

If we really want to go with pydantic it offers a special __root__ field for objects that are little more than a wrapper around a single data type ([here are the details](https://pydantic-docs.helpmanual.io/usage/models/#custom-root-types) in the pydantic documentation). So if we really wanted to use pydantic here, I think that would probably be the cleanest option. The disadvantage of that is that we would need to implement all of the dictionary methods that we might want by hand (`__getitem__`, `__setitem__`, etc...).

My proposal would be the following

```python

class RegionProcessor(Dict):

    def __setitem__(self, key, value):
        # protect against overwriting

    @classmethod
    def from_directory(cls, dirpath):
        # read all the region mappings from dirpath and set them as entries
        return cls

    def validate_against_dsd(self, dsd) -> None:
        # throw an error if something goes wrong else return None
    
    def aggregate_region(self, iam_df) -> IamDataFrame
        # perform the aggregation
```

In think that would be the cleanest solution, a slightly modified dictionary with protection against overwriting keys, a validation and an aggregation function.

Reply:
I still think it's not a good idea that the `RegionProcessor` is a dictionary subclass, and sorry that my suggestion about the private attribute sent you down a rabbit hole.

Alternative proposal:

```python
class RegionProcessor(BaseModel):
    mapping: Dict[str, RegionAggregationMapping]
...

@classmethod
    def from_directory(cls, path):
        # read all the region mappings from path and set them as entries
        # checking that no duplicates are added to `mapping`
        return cls(mapping)
```

So we only guard against duplicate model-mappings when parsing the directory, not in the actual class. A decent compromise between guarding against definition conflicts and using pydantic, with minimal re-implementation of method.

Reply:
No problem about the rabbit hole, it was a very educational one. 
I'm afraid I still don't fully understand the advantage of using pydantic in this case or the disadvantage of subclassing a dictionary. Only guarding against duplicate model mappings at initialization seems potentially a bit dangerous to me. If you think that's fine for our use case I will of course implement it using the `BaseModel` from pydantic.

Reply:
The advantage of using pydantic: if we use it for many classes, we should use it for all classes where there is no substantial argument against it.

The disadvantage of subclassing a dictionary (as I wrote above): the main use case of the `RegionProcessor is executing the processing, not just holding the collection of region mappings - that is just a pre-requisite for the main use case, but shouldn't drive the design decision.

Only guarding against duplicate mappings: there is a trade-off between a simple solution (which does not guard against all eventualities) and a more complex solution (which is more difficult to develop, test and maintain). In practice, 99% of the usage of the `RegionProcessor` will be just "import from directory, apply to an IamDataFrame". And guarding against duplicates only as part of the initialization from a directory is sufficient to test that project directories are consistent (ie, can be parsed, (target) regions exist in the `DataStructureDefinition`, no duplicate model mappings, internally consistent `RegionAggregationMapping`).

Reply:
Ok sounds good I'll implement it that way. Let's keep it consistent and not go overkill on making it 100% safe, which is impossible anyway.

I just realized that we need information from the `DataStructureDefinition` about how to execute the aggregation (simple sum, weighted average, etc...). For that it makes total sense to include that into the `RegionProcessor` and see it as a combination of a dict of mappings plus a `DataStructureDefinition`. My conceptual mistake was thinking that we only need the `DataStructureDefinition` to verify that all specified regions are allowed but we need it for more than that. That kills the dict approach. Sorry for the confusion.